File: ArchitecturePatternsExample/ArchitecturePatternsExample/ViewController.m
//
//  ViewController.m
//  ArchitecturePatternsExample
//
//  Created by ShannonChen on 2017/9/10.
//  Copyright © 2017年 ShannonChen. All rights reserved.
//

#import "ViewController.h"
#import "SCUserSelfProfileViewController.h"

@interface ViewController ()

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    
    
}


/**
 
 大体架构：
 
 - UserVC 作为业务场景 Scene, 需要展示三种数据, 对应的就有三个 MVC
 - 这三个 MVC 负责各自模块的数据获取, 数据处理和数据展示
 - UserVC 需要做的就是配置好这三个 MVC, 并在合适的时机通知各自的 C 层进行数据获取
 - 各个 C 层拿到数据后进行相应处理, 处理完成后渲染到各自的 View 上, UserVC 最后将已经渲染好的各个View 进行布局
 
 
 
 **/
    
- (IBAction)pushMVCExampleForSelfScene:(id)sender {
    [self.navigationController pushViewController:[[SCUserSelfProfileViewController alloc] init] animated:YES];
}
    
- (IBAction)pushMVCExampleForOtherScene:(id)sender {
    [self.navigationController pushViewController:[[SCUserProfileViewController alloc] initWithUserId:@"234"] animated:YES];
}


@end

File: ContentModeExample/ContentModeExample/ViewController.swift
//
//  ViewController.swift
//  ContentModeExample
//
//  Created by ShannonChen on 2017/9/8.
//  Copyright © 2017年 ShannonChen. All rights reserved.
//

import UIKit



class ViewController: UIViewController, UIImagePickerControllerDelegate, UINavigationControllerDelegate {
    
    // MARK: Properties
    let kAssetsImageName = "assetsImage"
    @IBOutlet weak var imageView: UIImageView!
    @IBOutlet weak var customImageView: CustomDrawingImageView!
    
    
    var imageName: String? {
        didSet {
            if let imageName = imageName, imageName != kAssetsImageName {
                imageView.image = UIImage.init(named: imageName)
//                customImageView.image = UIImage.init(named: imageName)
            }
        }
    }
    var contentMode: UIViewContentMode? {
        didSet {

            if let contentMode = contentMode {
                imageView.contentMode = contentMode
            }
            
        }
    }
    
    var clipToBounds: Bool? {
        didSet {
            if let clipToBounds = clipToBounds {
                imageView.clipsToBounds = clipToBounds
            }
        }
    }
    
//    var imageAlignment: UIImageViewAlignmentMask? {
//        didSet {
//            if let imageAlignment = imageAlignment {
//                imageView.alignment = imageAlignment
//            }
//            
//        }
//    }
    

    // MARK: Life Cycle
    override func viewDidLoad() {
        super.viewDidLoad()
        
        imageName = "inspiration_01"
        contentMode = UIViewContentMode.scaleAspectFill
        clipToBounds = true
    }
    
    override func viewDidLayoutSubviews() {
        super.viewDidLayoutSubviews()
        
        let imageWH = 20 / 375.0 * UIScreen.main.bounds.size.width
        
        imageView.frame.size = CGSize.init(width: imageWH, height: imageWH)
        imageView.center = view.center

    }
    
    
    // MARK: Actions
    @IBAction func selectImageFromPhotoLibrary(_ sender: UITapGestureRecognizer) {
        
        
        // UIIMagePickerController is a view controller that lets a user pick media from their photo library.
        let imagePickController = UIImagePickerController()
        
        // Only allow photos to be taken, not taken.
        imagePickController.sourceType = .photoLibrary
        
        // Make sure ViewController is notified when the user picks an image
        imagePickController.delegate = self
        
        // Present the picker
        present(imagePickController, animated: true, completion: nil)
        
    }
    
    // MARK: UIImagePickerControllerDelegate
    func imagePickerControllerDidCancel(_ picker: UIImagePickerController) {
        
        // Dismiss the picker if the user cancelled
        dismiss(animated: true, completion: nil)
    }
    
    func imagePickerController(_ picker: UIImagePickerController, didFinishPickingMediaWithInfo info: [String : Any]) {
        
        // The info dictionary contains multiple representaions of the image, and this uses the oringnal.
        let selectedImage = info[UIImagePickerControllerOriginalImage] as! UIImage
        
        // set photoImageView to display the selected image
        imageView.image = selectedImage
//        customImageView.image = selectedImage
        
        imageName = kAssetsImageName
        
        // Dismiss the picker
        dismiss(animated: true, completion: nil)
    }

}

